python/ray/dag/dag_node_operation.py

```diff
@@ -174,33 +234,60 @@ def _select_next_nodes(
         A list of _DAGOperationGraphNodes to be placed into the corresponding
         execution schedules.
     """
+    # If a node is a NCCL collective node, it updates the `ready_collective_nodes`
+    # of all the nodes in its collective group. It is removed from the candidate
+    # list if the collective group is not ready.
+    for actor, candidates in actor_to_candidates.items():
+        ready_candidates: List[_DAGOperationGraphNode] = []
+        for node in candidates:
+            if node.is_nccl_collective:
+                for collective_node_metadata in node.collective_group_idxs:
+                    task_idx, op_type = collective_node_metadata
+                    collective_node = graph[task_idx][op_type]
+                    collective_node.ready_collective_idxs.add(
+                        (node.task_idx, node.operation.type)
+                    )
+            if node.is_ready:
+                ready_candidates.append(node)
```

Comment:
`ready_candidates` should be a heap, so is there a possible bug here where removing a node from `ready_candidates` can mess with the heap order?

Instead of trying to remove from the middle of the heap, it would be cleaner to merge this logic with the code [here](https://github.com/ray-project/ray/pull/47621/files#r1780156424), so we do not push onto the heap until we're sure that the collective op is ready. That way, we can also do all of the `actor_to_candidates` pushing in one place, and this method is only allowed to pop from the head of the queue.

Reply:
A quick fix is `heapq.heappush(ready_candidates, node)`.

Let me think about merging the logic.

Reply:
Done.